truffle/src/main/java/org/jruby/truffle/nodes/ext/BigDecimalNodes.java

```diff
@@ -999,39 +1006,91 @@ public boolean zeroSpecial(RubyBasicObject value) {
     @CoreMethod(names = "sign")
     public abstract static class SignNode extends BigDecimalCoreMethodNode {
 
+        final private ConditionProfile positive = ConditionProfile.createBinaryProfile();
+        @Child private IntegerCastNode negativeInfinitySign;
+        @Child private IntegerCastNode positiveInfinitySign;
+        @Child private IntegerCastNode negativeZeroSign;
+        @Child private IntegerCastNode positiveZeroSign;
+        @Child private IntegerCastNode negativeFiniteSign;
+        @Child private IntegerCastNode positiveFiniteSign;
+        @Child private IntegerCastNode nanSign;
+
         public SignNode(RubyContext context, SourceSection sourceSection) {
             super(context, sourceSection);
         }
 
+        protected IntegerCastNode createGetSignConstantNodeChain(String constantName) {
+            final RubyContext context = getContext();
+            final SourceSection sourceSection = getSourceSection();
+
+            return IntegerCastNodeGen.create(context, sourceSection,
+                    ToIntNodeGen.create(context, sourceSection,
+                            GetConstantNodeGen.create(context, sourceSection,
+                                    new LiteralNode(context, sourceSection, getContext().getCoreLibrary().getBigDecimalClass()),
+                                    new LiteralNode(context, sourceSection, constantName),
+                                    LookupConstantNodeGen.create(context, sourceSection, LexicalScope.NONE, null, null))));
+        }
+
         @Specialization(guards = {
                 "isNormal(value)",
                 "isNormalZero(value)"})
-        public int signNormalZero(RubyBasicObject value) {
-            return 1;
+        public int signNormalZero(VirtualFrame frame, RubyBasicObject value) {
+            if (positiveZeroSign == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                positiveZeroSign = insert(createGetSignConstantNodeChain("SIGN_POSITIVE_ZERO"));
+            }
```

Comment:
Having `private static final int SIGN_NEGATIVE_INFINITE = -3` in `BigDecimalNodes.java`, and have let's say a "setup" method / interface implementation / etc in the same file, doing like
`c = defineClass(BigDecimal); c.setConstant("SIGN_NEGATIVE_INFINITE",  SIGN_NEGATIVE_INFINITE)`.
It would solve the pre-loading problem and the name mismatch (`Truffle::BigDecimal`).